v2/tools/generator/internal/codegen/pipeline/state.go

```diff
@@ -16,10 +16,11 @@ import (
 
 // State is an immutable instance that captures the information being passed along the pipeline
 type State struct {
-	definitions     astmodel.TypeDefinitionSet // set of type definitions generated so far
-	conversionGraph *storage.ConversionGraph   // graph of transitions between packages in our conversion graph
-	stagesSeen      set.Set[string]            // set of ids of the stages already run
-	stagesExpected  map[string]set.Set[string] // set of ids of expected stages, each with a set of ids for the stages expecting them
+	definitions        astmodel.TypeDefinitionSet // set of type definitions generated so far
+	conversionGraph    *storage.ConversionGraph   // graph of transitions between packages in our conversion graph
+	exportedConfigMaps *ExportedTypeNameProperties
```

Comment:
Why do we need this here, when we didn't add secrets here earlier?

Reply:
The reason that secrets doesn't need this is that we're not actually codegenerating the implementation of gathering the secrets. Whereas for configmaps we are. So for secrets the pattern is:
1. Create the `OperatorSpec.Secrets` object.
2. The actual mapping between the `OperatorSpec.Secrets.<prop>` properties and their data source is defined by hand in an extension, so the code generator doesn't need to know what it is.

For configMaps the pattern is:
1. Create the `OperatorSpec.ConfigMaps` object.
2. Code generate an implementation of `KubernetesExporter` which automatically maps the user specified values in `OperatorSpec.ConfigMaps` onto a `v1.ConfigMap` and saves the corresponding property.

To accomplish item 2 for ConfigMaps, we need to remember which property corresponds to which `OperatorSpec.ConfigMaps` entry. So for example `OperatorSpec.ConfigMaps.PrincipalId` maps to the `.Status.PrincipalId` field. We need this information twice, once to generate the `OperatorSpec.ConfigMaps` type, and once to generate the implementation of the `KubernetesExporter` interface. Since I already had it, I stashed it into `Stage` and pulled it out later.

I suppose an alternative implementation would be to re-call the `getConfigMapProperties(defs, configuration, resource)` function on each resource definition and rebuild the mapping in the following stage. We have the information in the type graph + config to do that. It's a bit "risky" though because it does remove the direct linkage between the set of properties used in `OperatorSpec.ConfigMaps` and the set of properties that are autogenerated. So if for some reason stages were moved such that we found a different list for some reason the second time round, that would be bad (I'd have to panic/error in that case, which can't happen now because we're using exactly the same data source). In practice right now that's not a risk though because the two stages one after the other.

Thoughts?

Reply:
This makes sense. Stashing the information on the `State` of the pipeline makes sense for this - it's a similar argument to why the conversion graph is there.

Thanks for clarifying!